source/agora/crypto/Hash.d

```diff
@@ -120,31 +118,30 @@ public Hash hashFull (T) (scope const auto ref T record)
 }
 
 /// Ditto
-public void hashPart (T) (scope const auto ref T record, scope HashDg state)
-    /*pure*/ nothrow @nogc
+public void hashPart (T) (scope const auto ref T record, scope HashDg hasher)
+    /*pure*/ nothrow @nogc @trusted
 {
-    static if (is(typeof(T.init.computeHash(HashDg.init))))
-        record.computeHash(state);
-    else static if (__traits(compiles, () { const ubyte[] r = T.init[]; }))
-        state(record[]);
-
-    else static if (isNarrowString!T)
-        state(cast(const(ubyte[]))record);
-    else static if (is(immutable(T) == immutable(ubyte[])))
-        state(record);
-
-    else static if (is(T : E[], E))
-        foreach (ref r; record)
-            hashPart(r, state);
-
-    else static if (is(immutable(ubyte) == immutable(T)))
-        state((cast(ubyte*)&record)[0 .. ubyte.sizeof]);
-    else static if (isScalarType!T)
-        state(nativeToLittleEndian(record)[0 .. T.sizeof]);
-
+    static if (is(T == struct))
+    {
+        static if (is(typeof(T.init.computeHash(HashDg.init))))
+            record.computeHash(hasher);
+        else
+            foreach (const ref field; record.tupleof)
+                hashPart(field, hasher);
+    }
     else
-        foreach (const ref field; record.tupleof)
-            hashPart(field, state);
+    {
+        // Enable the serializePart to be called even though it is not `@nogc nothrow`
+        alias serializePartAttributes = void delegate() @safe nothrow @nogc;
+        void hashWithSerializer () @safe
+        {
+            // We hash after binary serialization to prevent hash collisions related to empty arrays
+            // See issue https://github.com/bpfkorea/agora/issues/1331
+            serializePart(record, hasher, CompactMode.Yes);
+        }
+        scope serializePartAttributes correctAttributes = cast(serializePartAttributes) &hashWithSerializer;
+        correctAttributes();
```

Comment:
```suggestion
        // Enable `serializePart` to be called even though it is not `@nogc nothrow`
        alias SerializePartAttributes = void delegate() @safe nothrow @nogc;
        // We hash after binary serialization to prevent hash collisions related to empty arrays
        // See issue https://github.com/bpfkorea/agora/issues/1331
        scope hashWithSerializer = () { serializePart(record, hasher, CompactMode.Yes);  }
        (cast(SerializePartAttributes)hashWithSerializer)();
```

Try this. If it doesn't work, we can also try an intermediate variable for the call:
```suggestion
        // Enable `serializePart` to be called even though it is not `@nogc nothrow`
        alias SerializePartAttributes = void delegate() @safe nothrow @nogc;
        // We hash after binary serialization to prevent hash collisions related to empty arrays
        // See issue https://github.com/bpfkorea/agora/issues/1331
        scope hashWithSerializer = () { serializePart(record, hasher, CompactMode.Yes);  }
        scope correctAttributes = cast(SerializePartAttributes)hashWithSerializer;
        correctAttributes();
```

Reply:
Thanks, the second version worked with a small change.